Attestation (Limit use of authenticators)

jonashendrickx opened the pull request:
The application settings page will contain a new section to manage your allowed authenticators. It will act as a whitelist, which if populated will only allow registration using one of these authenticators.

Upon clicking the button, it will forward you to the list of currently configured whitelist (or blacklist in future). Whether or not the list is empty, you can add more authenticators to the whitelist.
<img width="1309" alt="image" src="https://github.com/bitwarden/passwordless-server/assets/1045327/de950e10-2e62-41af-8d1a-303b23acbf22">

<img width="1309" alt="image" src="https://github.com/bitwarden/passwordless-server/assets/1045327/bea222f9-bc43-4d7e-a23e-b80ade6bdc4e">

Authenticators that are actively whitelisted, are highlighted in blue to match the styling. You are free to add or remove authenticators that may or may not have been selected. No duplicates will be added in the database or similarly cause unexpected behavior.

<img width="1821" alt="image" src="https://github.com/bitwarden/passwordless-server/assets/1045327/ae0e5a0e-3115-4357-9b65-abd6848cb02b">

<img width="1821" alt="image" src="https://github.com/bitwarden/passwordless-server/assets/1045327/a5ffd8c2-2610-4974-a5c1-7645f583417f">

Added endpoints:
- GET /mds/entries: sorted alphabetically and filtered by attestation type or certification status optionally.
- GET /mds/attestation-types: grouped and sorted ascending
- GET /mds/certification-statuses: grouped and sorted ascending
- GET /authenticators/list: will list the currently configured authenticators.
- POST /authenticators/whitelist: whitelists authenticators accepting an array of aaguid.
- DELETE /authenticators/delist: removes authenticators from either whitelist or blacklist.

Authenticators endpoints require the application features to have their 'AllowedAttestation' set to true or will return a 403 otherwise. All pages have been similarly protected.

The `Authenticators` table was added to the `Api` project. This will be used to hold the configured authenticators and any relevant metadata we could require in the future.

`POST /register/complete` was impacted with the change, including already the integration for the blacklist if we decide to move forward with a blacklist at one point in the future.

jonashendrickx commented on src/AdminConsole/package.json:
```diff
@@ -11,10 +11,14 @@
+    "tailwindcss": "^3.4.0",
     "vue": "3.4.10"
+  },
+  "devDependencies": {
```
Extra dependencies for the Tailwind UI components.

jonashendrickx commented on src/AdminConsole/tailwind.config.js:
```diff
@@ -26,5 +26,9 @@ module.exports = {
         },
     },
-    plugins: ['@tailwindcss/forms'],
+    plugins: [
```
Tailwind Forms dependency was never working.

jonashendrickx commented on src/AdminConsole/Helpers/UseCsp.cs:
```diff
@@ -27,7 +27,8 @@ public static IApplicationBuilder UseCSP(this IApplicationBuilder app)
                 $"connect-src 'self' {passConfig.Value.ApiUrl};" +
-                "style-src 'self' 'unsafe-inline';";
+                "style-src 'self' 'unsafe-inline';" +
+                "img-src 'self' data:;";
```
For the dropdown icon to show.

abergs commented on src/Api/Endpoints/Apps.cs:
```diff
@@ -136,6 +137,64 @@ public static void MapAccountEndpoints(this WebApplication app)
+
+        app.MapGet("/apps/list-authenticators", ListConfiguredAuthenticatorsAsync)
+            .RequireSecretKey()
+            .RequireCors();
```
issue: This breaks the existing pattern of using `/list`. Is there an argument to be made to stray away from the convention?

suggestion: change this api pattern to follow convention`/apps/authenticators/list`

thought: I think we should do a team call to discuss API design for the `authenticators`-API. Can you schedule it?

abergs commented on src/Api/Endpoints/Apps.cs:
```diff
@@ -136,6 +137,64 @@ public static void MapAccountEndpoints(this WebApplication app)
+        app.MapDelete("/apps/delist-authenticators", DelistAuthenticatorsAsync)
+            .WithParameterValidation()
+            .RequireSecretKey()
+            .RequireCors();
```
thought: For API Design, I think we should `/authenticators`-API

abergs commented on src/Api/Endpoints/MetaDataService.cs:
```diff
@@ -0,0 +1,53 @@
+        var parent = app
+            .MapGroup("/mds")
+            .RequireSecretKey()
+            .RequireCors("default");
```
praise: Nice use of groups
nitpick: renam `parent` => `mds`.

thought: we should do a team API review for this as well. If these are assumed to be internal, we might need to stark indicate that in code, for us to keep track and in the future to avoid listing them in OpenAPI specs.

abergs reviewed (changes requested):
We discovered a weird UI bug that needs to be resolved:

https://files.slack.com/files-pri/T05CLP84TK2-F06FC4A12NP/cleanshot_2024-01-24_at_11.18.47.png

abergs commented on src/Service/SharedManagementService.cs:
```diff
@@ -409,6 +412,25 @@ public Task DisableGenerateSignInTokenEndpoint(string appId)
+    {
+        var storage = tenantFactory.Create();
+        return storage.DelistAuthenticatorsAsync(request.AaGuids);
+    }
```
issue: if Authenticators are scoped "per tenant" and/or always modified in an app-context, they should not live in the SharedManagementService. SharedManagementService is for non-app/tenant specific scenarios.

sggestion: Since it seems like these Authenticators do not do any business-logic, it seems it does not warrant its own service but could instead live in `ApplicationService`?

abergs commented on src/Service/Storage/Ef/DbGlobalContext.cs:
```diff
@@ -68,6 +69,16 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+                .WithMany(x => x.Authenticators)
+                .HasForeignKey(x => x.Tenant)
+                .IsRequired();
+        });
```
issue: This DBSet is missing the query filtering in `EFTenantContext` so any queries would impact all apps, not just the specific tenant.

abergs commented on src/Api/Endpoints/Authenticators.cs:
```diff
@@ -0,0 +1,74 @@
+{
+    public static void MapAuthenticatorsEndpoints(this WebApplication app)
+    {
+        var group = app.MapGroup("/mds")
```
issue: I think this is the wrong endpoint?

Tyrrrz commented on src/AdminConsole/Components/Shared/Page.cs:
```diff
@@ -0,0 +1,37 @@
+        builder.AddAttribute(1, nameof(PageTitle.ChildContent), (RenderFragment)(b =>
+        {
+            b.AddContent(2, Title);
+        }
+            ));
```
```suggestion
        }));
```

Tyrrrz commented on src/AdminConsole/Pages/App/Onboarding/GetStarted.cshtml:
```diff
@@ -1,23 +1,24 @@
     <div class="flex flex-col space-y-4 mt-2 max-w-xl text-sm text-gray-500">
         <p>Let's get set up using Passwordless.dev. Please feel free to read all the <a target="_blank" href="https://docs.passwordless.dev" class="link-blue">documentation</a>.</p>
-
+    
```
Remove spaces

```suggestion

```

Tyrrrz commented on src/AdminConsole/Services/PasswordlessManagement/IPasswordlessManagementClient.cs:
```diff
@@ -20,4 +21,23 @@ public interface IPasswordlessManagementClient
+    /// </summary>
+    /// <param name="request"></param>
+    /// <returns></returns>
+    Task<IEnumerable<EntryResponse>> GetMetaDataStatementEntriesAsync(EntriesRequest request);
```
Recommendation: return `Task<IReadOnlyCollection<EntryResponse>>` or `Task<IReadOnlyList<EntryResponse>>` so it's easier to work with the data (no double enumeration warnings, etc).

Tyrrrz commented on src/AdminConsole/Services/PasswordlessManagement/PasswordlessManagementClient.cs:
```diff
@@ -156,4 +158,37 @@ public async Task DisabledManuallyGeneratedTokensAsync(string appId, string perf
+
+    public async Task<IEnumerable<string>> GetAttestationTypesAsync()
+    {
+        var response = await http.GetAsync("/mds/attestation-types");
```
```suggestion
        using var response = await http.GetAsync("/mds/attestation-types");
        response.EnsureSuccessStatusCode();
```

Tyrrrz commented on src/AdminConsole/Services/PasswordlessManagement/PasswordlessManagementClient.cs:
```diff
@@ -156,4 +158,37 @@ public async Task DisabledManuallyGeneratedTokensAsync(string appId, string perf
+
+    public async Task<IEnumerable<string>> GetCertificationStatusesAsync()
+    {
+        var response = await http.GetAsync("/mds/certification-statuses");
```
```suggestion
        using var response = await http.GetAsync("/mds/certification-statuses");
        response.EnsureSuccessStatusCode();
```

Tyrrrz commented on src/AdminConsole/Services/ScopedPasswordlessClient.cs:
```diff
@@ -63,4 +85,24 @@ public async Task<IEnumerable<PeriodicCredentialReportResponse>> GetPeriodicCred
+
+    public async Task WhitelistAuthenticatorsAsync(WhitelistAuthenticatorsRequest request)
+    {
+        var response = await _client.PostAsJsonAsync("/authenticators/whitelist", request);
```
```suggestion
        using var response = await _client.PostAsJsonAsync("/authenticators/whitelist", request);
```

Tyrrrz commented on src/AdminConsole/Services/ScopedPasswordlessClient.cs:
```diff
@@ -63,4 +85,24 @@ public async Task<IEnumerable<PeriodicCredentialReportResponse>> GetPeriodicCred
+
+    public async Task DelistAuthenticatorsAsync(DelistAuthenticatorsRequest request)
+    {
+        var response = await _client.PostAsJsonAsync("/authenticators/delist", request);
```
```suggestion
        using var response = await _client.PostAsJsonAsync("/authenticators/delist", request);
```

Tyrrrz commented on src/AdminConsole/package.json:
```diff
@@ -11,10 +11,14 @@
     "apexcharts": "3.45.1",
     "es-module-shims": "1.8.2",
-    "tailwindcss": "3.4.0",
+    "tailwindcss": "^3.4.0",
     "vue": "3.4.10"
+  },
+  "devDependencies": {
+    "@tailwindcss/aspect-ratio": "^0.4.2",
+    "@tailwindcss/forms": "^0.5.7",
+    "@tailwindcss/typography": "^0.5.10"
```
Renovate suggests pinning dependencies:

```suggestion
    "tailwindcss": "3.4.0",
    "vue": "3.4.10"
  },
  "devDependencies": {
    "@tailwindcss/aspect-ratio": "0.4.2",
    "@tailwindcss/forms": "0.5.7",
    "@tailwindcss/typography": "0.5.10"
```

Tyrrrz commented on src/Api/Endpoints/Authenticators.cs:
```diff
@@ -0,0 +1,73 @@
+            .RequireSecretKey();
+    }
+
+    public static async Task<IResult> ListConfiguredAuthenticatorsAsync(
```
Nit: I'd personally inline the methods into delegates, both for consistency and because they're quite small anyway

Tyrrrz commented on src/Api/Endpoints/Authenticators.cs:
```diff
@@ -0,0 +1,73 @@
+            .RequireCors("default");
+
+        group.MapGet("/list", ListConfiguredAuthenticatorsAsync)
+            .RequireSecretKey();
```
Put `.RequireSecretKey()` on the group instead?

Tyrrrz commented on src/Common/Validation/RegularExpressionCollectionAttribute.cs:
```diff
@@ -0,0 +1,31 @@
+
+        if (value is not IEnumerable<string> collection)
+        {
+            throw new InvalidOperationException("This attribute can only be applied to collections");
```
Nit: dots are customary at the end of exception messages

```suggestion
            throw new InvalidOperationException("This attribute can only be applied to collections.");
```

Tyrrrz commented on tests/Api.Tests/Endpoints/AuthenticatorsEndpointsTests.cs:
```diff
@@ -0,0 +1,201 @@
+
+    #region ListConfiguredAuthenticatorsAsync
+    [Fact]
+    public async Task ListConfiguredAuthenticatorsAsync_Throws_Forbidden_WhenAttestationIsNotAllowed()
```
Can we perform these tests at an integration level instead?

jonashendrickx replied:
@Tyrrrz Yeah, but I wanted to refactor the integration tests first, so less steps would be required to create an app, etc.

jonashendrickx replied:
@Tyrrrz  done

Tyrrrz commented on tests/Api.Tests/Endpoints/AuthenticatorsEndpointsTests.cs:
```diff
@@ -0,0 +1,201 @@
+    public async Task ListConfiguredAuthenticatorsAsync_Throws_Forbidden_WhenAttestationIsNotAllowed()
+    {
+        // Arrange
+        var request = _fixture.Create<ConfiguredAuthenticatorRequest>();
```
Are these tests needed now that `AuthenticatorsTests` are in place? Do they cover something else?

jonashendrickx replied:
Yeah, I use them to quickly verify some expected behavior

abergs commented on src/Common/Validation/RegularExpressionCollectionAttribute.cs:
```diff
@@ -0,0 +1,31 @@
+namespace Passwordless.Common.Validation;
+
+[AttributeUsage(AttributeTargets.Property)]
+public class RegularExpressionCollectionAttribute(string pattern) : RegularExpressionAttribute(pattern)
```
suggestion: Add /// comments to explain what this is

abergs commented on src/Api/Endpoints/Authenticators.cs:
```diff
@@ -0,0 +1,76 @@
+            .WithParameterValidation();
+
+        group.MapPost("/delist", DelistAuthenticatorsAsync)
+            .WithParameterValidation();
```
issue: I'd like to find a better design for this api. I like the `/authenticators/` namespace, but I don't like whitelist and delist.

What about `/add` {isAllowed: true} and a `/remove`.

Since you technically added support for the blocklist in the db, this would just reflect that.

abergs commented:
Exception: While running on my machine, I navigate to the Settings > Authenticators in Admin Console. The API Crashes on this exception:

```
System.IO.DirectoryNotFoundException: Could not find a part of the path '/Users/anders/Github/passwordless/passwordless-server/src/Api/.mds-cache/mds.jwt'.
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, UnixFileMode openPermissions, Int64& fileLength, UnixFileMode& filePermissions, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
   at System.IO.StreamWriter.ValidateArgsAndOpenPath(String path, Boolean append, Encoding encoding, Int32 bufferSize)
   at System.IO.File.CreateText(String path)
   at Passwordless.Service.MDS.CacheHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /Users/anders/Github/passwordless/passwordless-server/src/Service/MDS/CacheHandler.cs:line 29
```

abergs commented:
A couple of UI suggestions:

**suggestion**: Add a way to navigate back to previous place in our UI, e.g. a breadcrum.
![CleanShot 2024-01-25 at 10 27 01](https://github.com/bitwarden/passwordless-server/assets/357283/032ad190-0608-4982-aad8-218cc30f8e5e)

**suggestion**: Remove Search and Reset buttons when filtering, just allow me to make the selection and automatically search.
![CleanShot 2024-01-25 at 10 27 49](https://github.com/bitwarden/passwordless-server/assets/357283/a292c9e8-e908-4de6-b4b0-68f82b245521)

**suggestion**: Add a more distinct way to visualise that the authenticator is Allowed. In the image below, I was confused which was allowed and which wasn't.
![CleanShot 2024-01-25 at 10 37 15](https://github.com/bitwarden/passwordless-server/assets/357283/b87ac715-cc1b-41b2-847d-e5114c167305)

To the point, I'm wondering if we can remove the "Add authenticator/Remove authenticator/clear" buttons and just use the checkbox. Name the checkboxbox column "Allowed". When toggled on, it's added to the allow list, when toggled off it's removed. The UI would be simplified to this:

**Updated UI:** 
![CleanShot 2024-01-25 at 10 36 29](https://github.com/bitwarden/passwordless-server/assets/357283/71c9a815-fa6a-4520-a873-78c8f0ada8fe)

abergs commented:
This might be irrelevant if we change the UI interaction.

When I click either the Add Authenticators or Remove Authenticators buttons, without selecting any checkboxes, the page crashes:
![CleanShot 2024-01-25 at 10 41 51](https://github.com/bitwarden/passwordless-server/assets/357283/970bd8cc-cd7c-4282-8dea-d0da7cca30fe)

Result:

![CleanShot 2024-01-25 at 10 40 34](https://github.com/bitwarden/passwordless-server/assets/357283/ee5bb4d2-826b-4129-9b81-e01be45feebc)

abergs commented on src/AdminConsole/Components/Pages/App/BaseApplicationPage.cs:
```diff
@@ -0,0 +1,12 @@
+    /// The route parameter for the application id for application scoped pages.
+    /// </summary>
+    [Parameter]
+    public required string App { get; set; }
```
@jonashendrickx Just realised, this should probably be AppId for consistency sake

jonashendrickx replied:
@abergs No. It will not work. This binds to the 'app' route parameter. If I rename it, it will also break all the other middleware.

abergs replied:
Aha, I see. We should probably refactor that in a separate issue. I'll create it.

abergs commented on src/AdminConsole/Components/Pages/App/Reporting.razor:
```diff
@@ -1,24 +1,15 @@
 @page "/app/{app}/Reporting"
```
issue: With the move from ´App´ to AppId, this pattern needs to change too.

jonashendrickx commented on tests/AdminConsole.Tests/Components/Pages/App/BaseApplicationPageTests.cs:
```diff
@@ -0,0 +1,74 @@
+/// <summary>
+/// This is verifying that all application pages inherit from <c href="BaseApplicationPage" /> and that <c href="HasAppHandler" /> authorization policy is applied.
+/// </summary>
+public class BaseApplicationTests
```
Policy enforcement tests

jonashendrickx commented on src/AdminConsole/Components/Pages/App/BaseApplicationPage.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Passwordless.AdminConsole.Components.Pages.App;
+
+[Authorize(CustomPolicy.HasAppRole)]
```
Applying the policy to all application pages

jonashendrickx commented on src/AdminConsole/Components/Routes.razor:
```diff
@@ -1,13 +1,18 @@
-        <AuthorizeRouteView RouteData="@routeData" DefaultLayout="@typeof(Layouts.MainLayout)">
+        <!-- The 'Resource' property will be used in 'HasAppHandler' to obtain the 'app' route parameter -->
+        <!-- If we implement new policies to protect resources, we should protect them with resource based uri and compare them with the user's claims. -->
+        <AuthorizeRouteView RouteData="@routeData" Resource="@HttpContextAccessor.HttpContext" DefaultLayout="@typeof(Layouts.MainLayout)">
```
`HasAppHandler` reads the `app` route parameter from the HttpContext. This strategy will leave us with maximum flexibility for new policies.

abergs commented:
Good work with the improvements @jonashendrickx, things are definitely getting there!

Missing some things from my [previous feedback](https://github.com/bitwarden/passwordless-server/pull/372#issuecomment-1909760380), i'll re-iterate them here:

There is no way to navigate back to Allowed Authenticators, suggestion is to add a breadcrumb-style back-link.
![CleanShot 2024-01-25 at 10 27 01](https://github.com/bitwarden/passwordless-server/assets/357283/580d6dcc-7fc6-454d-8b40-aceafbf51923)

Missing column-name for checkboxes:

![CleanShot 2024-01-31 at 19 38 42](https://github.com/bitwarden/passwordless-server/assets/357283/1085297e-6e36-4687-877c-03ca2ed3de89)

We could improve our message on this screen.

![CleanShot 2024-01-31 at 19 40 34](https://github.com/bitwarden/passwordless-server/assets/357283/0ca9bcdc-33b7-4dc6-8386-89df607c47ac)

> If this list is empty, all authenticators are allowed. Use this list if you want to control what authenticators can be used to register credentials.

I don't think it's obvious there's a list here at all when it's empty, I'd suggest adding a different text when the list is in fact empty. Suggestion:

**All authenticators are allowed by default**. If you wish, you can control what authenticators are allowed. Button text: Choose Allowed authenticators...

And perhaps change the non-empty text to:

**Only the authenticators listed below are allowed**. If you wish you can allow any authenticator or change your selection. Button text: Choose Allowed authenticators...

abergs reviewed (changes requested):
See previous comment.

jonashendrickx commented:
<img width="2056" alt="image" src="https://github.com/bitwarden/passwordless-server/assets/1045327/70756d82-d441-4ed5-bd49-6f945914dcca">

<img width="2056" alt="image" src="https://github.com/bitwarden/passwordless-server/assets/1045327/5fb173c2-4205-49d8-878b-23f9b86f1a47">

The last link is not clickable. If the settings page is migrated to Blazor, we can automatically resolve all parents using reflection.

jonashendrickx commented:
<img width="2056" alt="image" src="https://github.com/bitwarden/passwordless-server/assets/1045327/a01d5eea-77d2-4112-94a1-28b05f5ba218">

Changed my mind, underlined the currently viewing page, and removed the hover effect for it given it's not interactable.
